src/libfsm/endids.c

```diff
@@ -270,6 +270,72 @@ endid_find_bucket(struct fsm *fsm, fsm_state_t state)
 	return NULL;
 }
 
+/* Returns the entry k of ids such that:
+ * 0 <= k <= n
+ * if 0 < k < n, then: ids[k-1] < itm <= ids[k]
+ * if k == 0, then: itm <= ids[0]
+ * if k == n, then: ids[n-1] < itm
+ *
+ * Note that ids must be sorted in ascending order: if i <= j, then ids[i] <= ids[j]
```

Comment:
That's probably a good idea.  I do have a check for ordering in `fsm_endid_set()`:
```
DBG_3(assert(check_ids_sorted(b->ids->ids, b->ids->count)));
```

but putting it under a more explicit `EXPENSIVE_CHECKS`-ish thing instead of `LOG_ENDIDS>3` makes sense.

> (I would probably spend a lot of time triple-checking that, because I never get binary search right the first time.)

Yes, this is one of my concerns.  I would like to have a unit test for the binary search itself, but there isn't a clear way to unit test internal apis for `libfsm` right now.  Moving some of the `fsm_end_id_t` code into `adt/` might be one way, but it's a lot for a fairly specific bit of code.

Reply:
Changed `DBG_3(assert(check_ids_sorted(...))` to `assert(check_ids_sorted(...))` and added `ENDIDS_EXPENSIVE_CHECK_BISECTION` guard to disable the check in [1379dea](https://github.com/katef/libfsm/pull/426/commits/1379dea00ee8f11e16e350a03d615f67227d1401).

Added a similar check to the binary search in [280c65d](https://github.com/katef/libfsm/pull/426/commits/280c65d62496d3693264d7faa632b7458ecacdd9)

Reply:
if we're having more granular "expensive checks" macros, let's add them to CI too (i already included `EXPENSIVE_CHECKS` there). personally i don't think this granularity gives us any advantage, and it's easily missed with different macros like this. i'd rather we just have one. although to be honest, i'd rather we just have `NDEBUG` for all cases, but that's a case i'll make another day